README
NAME
    String::Dump - Dump strings of characters (or bytes) for printing and
    debugging

VERSION
    This document describes String::Dump version 0.08.

SYNOPSIS
        use String::Dump qw( dump_hex dump_bin );

        say 'hex: ', dump_hex($string);
        say 'bin: ', dump_bin($string);

DESCRIPTION
    When debugging or examining strings containing non-ASCII or non-printing
    characters, String::Dump is your friend. It provides simple functions to
    return a dump of the code points for Unicode strings or the bytes for
    byte strings in several different formats, such as hex, binary, Unicode
    names, and more.

    For using this module from the command line, see the bundled dumpstr
    script. For tips on debugging Unicode or byte strings with this module,
    see the document String::Dump::Debugging.

FUNCTIONS
    These functions all accept a single argument: the string to dump, which
    may either be a Unicode string or a byte string. Each function has to be
    explicitly exported or they can all be exported with the ":all" tag.

        use String::Dump qw( :all );

  dump_hex($string)
    Hexadecimal (base 16) mode.

        use utf8;
        # string of 6 characters
        say dump_hex('Ĝis! ☺');  # 11C 69 73 21 20 263A

        no utf8;
        # series of 9 bytes
        say dump_hex('Ĝis! ☺');  # C4 9C 69 73 21 20 E2 98 BA

    For a lowercase hex dump, simply pass the response to "lc".

        say lc dump_hex('Ĝis! ☺');  # 11c 69 73 21 20 263a

  dump_dec($string)
    Decimal (base 10) mode. This is mainly useful when referencing 8-bit
    code pages like ISO-8859-1 or 7-bit ones like ASCII variants.

        use utf8;
        say dump_dec('Ĝis! ☺');  # 284 105 115 33 32 9786

        no utf8;
        say dump_dec('Ĝis! ☺');  # 196 156 105 115 33 32 226 152 186

  dump_oct($string)
    Octal (base 8) mode. This is mainly useful when referencing 7-bit code
    pages like ASCII variants.

        use utf8;
        say dump_oct('Ĝis! ☺');  # 434 151 163 41 40 23072

        no utf8;
        say dump_oct('Ĝis! ☺');  # 304 234 151 163 41 40 342 230 272

  dump_bin($string)
    Binary (base 2) mode.

        use utf8;
        say dump_bin('Ĝis! ☺');
        # 100011100 1101001 1110011 100001 100000 10011000111010

        no utf8;
        say dump_bin('Ĝis! ☺');
        # 11000100 10011100 1101001 1110011 100001 100000 11100010 10011000 10111010

  dump_names($string)
    Unicode character name mode. Unlike the various numeral modes above,
    this mode uses “, ” <comma, space> for the delimiter and it only makes
    sense for Unicode strings, not byte strings.

        use utf8;
        say dump_names('Ĝis! ☺');
        # LATIN CAPITAL LETTER G WITH CIRCUMFLEX, LATIN SMALL LETTER I,
        # LATIN SMALL LETTER S, EXCLAMATION MARK, SPACE, WHITE SMILING FACE

    The output in the example above has been manually split into multiple
    lines for the layout of this document.

  dump_codes($string)
    Unicode code point mode. This is similar to "dump_hex" except it follows
    the standard Unicode code point notation. The hex value is 4 to 6
    digits, padded with “0” <digit zero> when less than 4, and prefixed with
    “U+” <latin capital letter u, plus sign>. As with "dump_names", this
    function only makes sense for Unicode strings, not byte strings.

        use utf8;
        say dump_codes('Ĝis! ☺');  # U+011C U+0069 U+0073 U+0021 U+0020 U+263A

SEE ALSO
    *   dumpstr - Dump strings of characters on the command line

    *   String::Dump::Debugging - String debugging tips with String::Dump

    *   Template::Plugin::StringDump - String::Dump plugin for TT

    *   Data::HexDump - Simple hex dumping using the default output of the
        Unix "hexdump" utility

    *   Data::Hexdumper - Advanced formatting of binary data, similar to
        "hexdump"

AUTHOR
    Nick Patch <patch@cpan.org>

COPYRIGHT AND LICENSE
    © 2011–2012 Nick Patch

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl module String::Dump

0.08 2012-10-24
 - Added dump_codes()
 - Removed functions from automatic exports after deprecating in 0.06
 - Converted this file from YAML to the standard CPAN Changes format
 - Improved docs

0.07 2012-03-04
 - Simplify and optimize algorithms
 - Minor documentation updates

0.06 2012-02-19
 - Removed dumpstr() and dump_string()
 - NOTE: Deprecated implicit exporting of functions

0.05 2011-12-27
 - Added dump_hex(), dump_dec(), dump_oct(), dump_bin(), dump_names() functions
 - Added dumpstr cammand-line tool
 - Moved POD tips to String::Dump::Debugging
 - NOTE: dumpstr() and dump_string() will be removed in the next release
 - APOLOGY: Sorry for all the recent design changes :P

0.04 2011-12-14
 - Replaced unknown Unicode names in names mode with '?'
 - Added parameter validations
 - Removed dump_string from automatic exports

0.03 2011-12-12
 - Renamed dump_string() to dumpstr()
 - Added dump_string() as an alias of dumpstr()
 - NOTE: Automatic dump_string() exporting will be removed in the next release

0.02 2011-12-12
 - Improved POD, Build.PL, and INSTALL

0.01 2011-12-10
 - First release


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name    => 'String::Dump',
    license        => 'perl',
    create_license => 1,
    create_readme  => 1,
    configure_requires => {
        'Module::Build' => 0.36,
    },
    build_requires => {
        'Test::More' => 0.82,  # note()
        'Test::Warn' => 0,
    },
    requires => {
        'perl'   => '5.6.0',  # sprintf('%b'), charnames::viacode(), our()
        'parent' => 0,
    },
    meta_merge => {
        resources => {
            repository => 'git://github.com/patch/string-dump-pm5.git',
            homepage   => 'https://github.com/patch/string-dump-pm5',
            bugtracker => 'https://github.com/patch/string-dump-pm5/issues',
        },
    },
);

$builder->create_build_script;


lib/String/Dumper.pm
package String::Dumper;

use 5.006;
use strict;
use warnings;
use charnames qw( :full );
use parent 'Exporter';

our $VERSION = '0.01';
our @EXPORT  = qw( dump_string );

use constant DEFAULT_MODE => 'hex';

my %delim_for = (
    hex   => ' ',
    dec   => ' ',
    oct   => ' ',
    bin   => ' ',
    names => ', ',
);

my %sub_for = (
    hex   => sub { map { sprintf '%X',      ord } @_ },
    dec   => sub { map {                    ord } @_ },
    oct   => sub { map { sprintf '%o',      ord } @_ },
    bin   => sub { map { sprintf '%b',      ord } @_ },
    names => sub { map { charnames::viacode ord } @_ },
);

sub dump_string {
    my ($mode, $string);

    if (@_ == 1) {
        ($mode, $string) = (DEFAULT_MODE, @_);
    }
    else {
        ($mode, $string) = @_;
    }

    return unless defined $string;

    return join $delim_for{$mode}, $sub_for{$mode}->(split '', $string);
}

1;

__END__

=encoding utf8

=head1 NAME

String::Dumper - Dump strings of characters or bytes for printing and debugging

=head1 VERSION

This document describes String::Dumper version 0.01.

=head1 SYNOPSIS

    use String::Dumper;

    say 'hex: ', dump_string($string);  # hex mode by default
    say 'oct: ', dump_string(oct => $string);  # octal mode

=head1 NOTICE

This is an early release of String::Dumper.  Feedback is appreciated!  To give
suggestions or report an issue, contact L<mailto:patch@cpan.org> or open an
issue at L<https://github.com/patch/string-dumper-pm5/issues>.  Pull requests
are welcome at L<https://github.com/patch/string-dumper-pm5>.

=head1 DESCRIPTION

This module provides the function C<dump_string> and exports it by default.
When debugging or reviewing non-ASCII data, C<dump_string> is your friend.
It's a simple utility to view the characters or bytes of your string in
several different formats, such as hex, octal, decimal, Unicode names, and
more.

An OO interface is forthcoming with additional options and the ability to
reuse them among multiple calls.  Some benefits will be the ability to set
the delimiter between characters and to force a string to be treated as a
string of characters or a series of bytes.  Don't worry, the C<dump_string>
function will remain simple.

=head2 dump_string($mode, $string)

The mode is optional and defaults to C<hex>.  Other valid modes are C<dec>,
C<oct>, C<bin>, and C<names>, and are described below.  The string may either
be a series of Unicode characters or binary bytes.

=head2 Modes

=head3 hex

Hexadecimal (base 16) mode.  This is the default when only a string is passed
without the mode.

    use utf8;
    # string of 6 characters
    say dump_string('Ĝis! ☺');  # 11C 69 73 21 20 263A
    say dump_string(hex => 'Ĝis! ☺');  # same thing

    no utf8;
    # series of 9 bytes
    say dump_string('Ĝis! ☺');  # C4 9C 69 73 21 20 E2 98 BA

For a lowercase hex dump, simply pass the response to C<lc>.

    say lc dump_string('Ĝis! ☺');  # 11c 69 73 21 20 263a

=head3 dec

Decimal (base 10) mode.

    use utf8;
    say dump_string(dec => 'Ĝis! ☺');  # 284 105 115 33 32 9786

    no utf8;
    say dump_string(dec => 'Ĝis! ☺');  # 196 156 105 115 33 32 226 152 186

=head3 oct

Octal (base 8) mode.

    use utf8;
    say dump_string(oct => 'Ĝis! ☺');  # 434 151 163 41 40 23072

    no utf8;
    say dump_string(oct => 'Ĝis! ☺');  # 304 234 151 163 41 40 342 230 272

=head3 bin

Binary (base 2) mode.

    use utf8;
    say dump_string(bin => 'Ĝis! ☺');
    # 100011100 1101001 1110011 100001 100000 10011000111010

    no utf8;
    say dump_string(bin => 'Ĝis! ☺');
    # 11000100 10011100 1101001 1110011 100001 100000 11100010 10011000 10111010

=head3 names

Named Unicode character mode.  Unlike the various numeral modes above, this
mode uses ', ' for the delimiter.

    use utf8;
    say dump_string(names => 'Ĝis! ☺');
    # LATIN CAPITAL LETTER G WITH CIRCUMFLEX, LATIN SMALL LETTER I,
    # LATIN SMALL LETTER S, EXCLAMATION MARK, SPACE, WHITE SMILING FACE

This mode makes no sense for a series of bytes, but it still works if that's
what you really want!

    no utf8;
    say dump_string(names => 'Ĝis! ☺');
    # LATIN CAPITAL LETTER A WITH DIAERESIS, STRING TERMINATOR,
    # LATIN SMALL LETTER I, LATIN SMALL LETTER S, EXCLAMATION MARK,
    # SPACE, LATIN SMALL LETTER A WITH CIRCUMFLEX, START OF STRING,
    # MASCULINE ORDINAL INDICATOR

The output in the examples above has been manually split into multiple lines
for the layout of this document.

=head2 Tips

=head3 Literal strings

When dumping literal strings in your code, as in the examples above, use the
L<utf8> pragma when strings of Unicode characters are desired and don't use it
or disable it when series of bytes are desired.  The pragma may also be
lexically enabled or disabled.

    use utf8;

    {
        no utf8;
        say dump_string('Ĝis! ☺');  # C4 9C 69 73 21 20 E2 98 BA
    }

    say dump_string('Ĝis! ☺');  # 11C 69 73 21 20 263A

=head3 Command-line input and filehandles

The simplest way to ensure that you're working with strings of characters from
all of your basic sources of input is to use the L<utf8::all> pragma.  This
extends the utf8 pragma to automatically convert command-line arguments
provided by C<@ARGV>, user-defined filehandles, as well as C<STDIN>, among
others.

=head3 Other sources of input

To handle strings provided by other sources of input, such as from network
protocols or a web server request, pass the value to
L<Encode::decode_utf8|Encode>, which will return the desired string.

    use Encode;

    say dump_string( decode_utf8($string) );

To convert a variable in-place, pass it to utf8::decode instead.

    utf8::decode($string);

    say dump_string($string);

=head1 SEE ALSO

=over

=item * L<Data::HexDump> - Simple hex dumping using the default output of the
Unix C<hexdump> utility

=item * L<Data::Hexdumper> - Advanced formatting of binary data, similar to
C<hexdump>

=back

=head1 AUTHOR

Nick Patch <patch@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright 2011 Nick Patch

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/String/Dump.pm
package String::Dump;

use 5.006;
use strict;
use warnings;
use parent 'Exporter';
use charnames qw( :full );
use Carp;

our $VERSION = '0.08';
our @EXPORT_OK = qw(
    dump_hex
    dump_dec
    dump_oct
    dump_bin
    dump_names
    dump_codes
);
our %EXPORT_TAGS = (all => \@EXPORT_OK);

sub dump_hex {
    my ($str) = @_;
    carp('dump_hex() expects one argument') && return if @_ != 1;
    return unless defined $str;
    return sprintf '%*vX', ' ', $str;
}

sub dump_dec {
    my ($str) = @_;
    carp('dump_dec() expects one argument') && return if @_ != 1;
    return unless defined $str;
    return sprintf '%*vd', ' ', $str;
}

sub dump_oct {
    my ($str) = @_;
    carp('dump_oct() expects one argument') && return if @_ != 1;
    return unless defined $str;
    return sprintf '%*vo', ' ', $str;
}

sub dump_bin {
    my ($str) = @_;
    carp('dump_bin() expects one argument') && return if @_ != 1;
    return unless defined $str;
    return sprintf '%*vb', ' ', $str;
}

sub dump_names {
    my ($str) = @_;
    carp('dump_names() expects one argument') && return if @_ != 1;
    return unless defined $str;
    return join ', ',
           map { charnames::viacode(ord) || '?' }
           split '', $str;
}

sub dump_codes {
    my ($str) = @_;
    carp('dump_codes() expects one argument') && return if @_ != 1;
    return unless defined $str;
    return join ' ', map { sprintf 'U+%04X', ord } split '', $str;
}

1;

__END__

=encoding UTF-8

=head1 NAME

String::Dump - Dump strings of characters (or bytes) for printing and debugging

=head1 VERSION

This document describes String::Dump version 0.08.

=head1 SYNOPSIS

    use String::Dump qw( dump_hex dump_bin );

    say 'hex: ', dump_hex($string);
    say 'bin: ', dump_bin($string);

=head1 DESCRIPTION

When debugging or examining strings containing non-ASCII or non-printing
characters, String::Dump is your friend.  It provides simple functions to return
a dump of the code points for Unicode strings or the bytes for byte strings in
several different formats, such as hex, binary, Unicode names, and more.

For using this module from the command line, see the bundled L<dumpstr> script.
For tips on debugging Unicode or byte strings with this module, see the document
L<String::Dump::Debugging>.

=head1 FUNCTIONS

These functions all accept a single argument: the string to dump, which may
either be a Unicode string or a byte string.  Each function has to be explicitly
exported or they can all be exported with the C<:all> tag.

    use String::Dump qw( :all );

=head2 dump_hex($string)

Hexadecimal (base 16) mode.

    use utf8;
    # string of 6 characters
    say dump_hex('Ĝis! ☺');  # 11C 69 73 21 20 263A

    no utf8;
    # series of 9 bytes
    say dump_hex('Ĝis! ☺');  # C4 9C 69 73 21 20 E2 98 BA

For a lowercase hex dump, simply pass the response to C<lc>.

    say lc dump_hex('Ĝis! ☺');  # 11c 69 73 21 20 263a

=head2 dump_dec($string)

Decimal (base 10) mode.  This is mainly useful when referencing 8-bit code pages
like ISO-8859-1 or 7-bit ones like ASCII variants.

    use utf8;
    say dump_dec('Ĝis! ☺');  # 284 105 115 33 32 9786

    no utf8;
    say dump_dec('Ĝis! ☺');  # 196 156 105 115 33 32 226 152 186

=head2 dump_oct($string)

Octal (base 8) mode.  This is mainly useful when referencing 7-bit code pages
like ASCII variants.

    use utf8;
    say dump_oct('Ĝis! ☺');  # 434 151 163 41 40 23072

    no utf8;
    say dump_oct('Ĝis! ☺');  # 304 234 151 163 41 40 342 230 272

=head2 dump_bin($string)

Binary (base 2) mode.

    use utf8;
    say dump_bin('Ĝis! ☺');
    # 100011100 1101001 1110011 100001 100000 10011000111010

    no utf8;
    say dump_bin('Ĝis! ☺');
    # 11000100 10011100 1101001 1110011 100001 100000 11100010 10011000 10111010

=head2 dump_names($string)

Unicode character name mode.  Unlike the various numeral modes above, this mode
uses “, ” <comma, space> for the delimiter and it only makes sense for Unicode
strings, not byte strings.

    use utf8;
    say dump_names('Ĝis! ☺');
    # LATIN CAPITAL LETTER G WITH CIRCUMFLEX, LATIN SMALL LETTER I,
    # LATIN SMALL LETTER S, EXCLAMATION MARK, SPACE, WHITE SMILING FACE

The output in the example above has been manually split into multiple lines for
the layout of this document.

=head2 dump_codes($string)

Unicode code point mode.  This is similar to C<dump_hex> except it follows the
standard Unicode code point notation.  The hex value is 4 to 6 digits, padded
with “0” <digit zero> when less than 4, and prefixed with “U+” <latin capital
letter u, plus sign>.  As with C<dump_names>, this function only makes sense for
Unicode strings, not byte strings.

    use utf8;
    say dump_codes('Ĝis! ☺');  # U+011C U+0069 U+0073 U+0021 U+0020 U+263A

=head1 SEE ALSO

=over

=item * L<dumpstr> - Dump strings of characters on the command line

=item * L<String::Dump::Debugging> - String debugging tips with String::Dump

=item * L<Template::Plugin::StringDump> - String::Dump plugin for TT

=item * L<Data::HexDump> - Simple hex dumping using the default output of the
Unix C<hexdump> utility

=item * L<Data::Hexdumper> - Advanced formatting of binary data, similar to
C<hexdump>

=back

=head1 AUTHOR

Nick Patch <patch@cpan.org>

=head1 COPYRIGHT AND LICENSE

© 2011–2012 Nick Patch

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut
